sklearn/preprocessing/_data.py

```diff
@@ -789,39 +799,32 @@ def partial_fit(self, X, y=None, sample_weight=None):
             sparse_constructor = (sparse.csr_matrix
                                   if X.format == 'csr' else sparse.csc_matrix)
 
-            if sample_weight is not None:
-                raise NotImplementedError(
-                    "`StandardScaler` does not yet have a support for sparse X"
-                    " and `sample_weight`.")
-            else:
-                counts_nan = sparse_constructor(
-                        (np.isnan(X.data), X.indices, X.indptr),
-                        shape=X.shape).sum(axis=0).A.ravel()
-
-            if not hasattr(self, 'n_samples_seen_'):
-                self.n_samples_seen_ = (
-                    X.shape[0] - counts_nan).astype(np.int64, copy=False)
-
             if self.with_std:
                 # First pass
                 if not hasattr(self, 'scale_'):
-                    self.mean_, self.var_ = mean_variance_axis(X, axis=0)
+                    self.mean_, self.var_, self.n_samples_seen_ = \
+                        mean_variance_axis(X, axis=0, weights=sample_weight,
+                                           return_sum_weights=True)
                 # Next passes
                 else:
                     self.mean_, self.var_, self.n_samples_seen_ = \
                         incr_mean_variance_axis(X, axis=0,
                                                 last_mean=self.mean_,
                                                 last_var=self.var_,
-                                                last_n=self.n_samples_seen_)
+                                                last_n=self.n_samples_seen_,
+                                                weights=sample_weight)
             else:
-                self.mean_ = None
+                self.mean_ = None  # as with_mean must be False for sparse
                 self.var_ = None
-                if hasattr(self, 'scale_'):
-                    self.n_samples_seen_ += X.shape[0] - counts_nan
+                if sample_weight is None:
+                    weights = np.ones(n_samples, dtype=np.int64)
+                else:
+                    weights = sample_weight
```

Comment:
We should use `sklearn.utils.validation._check_sample_weight`. It will manage this initialization indeed.